### PR TITLE
Add post re MyUW uPortal fork

### PR DESCRIPTION
**Pull request by apetro**
Add post describing MyUW's uPortal fork.

### CODE REVIEW

File: _posts/uportal/2017-12-06-myuw-fork.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+MyUW is running uPortal 4.2.1 with local customizations that both pre-date and post-date that release, the former achieved through heroic source control merging of a long history of local change.
+
+This was closer to a typical historical uPortal adoption model, building the 
+local portal from source. With ready access to that source, it's tempting to 
+enhance it. With long release cycles and difficult upgrades, making 
+enhancements directly in upstream often felt like duplicated effort, since to 
+realize the improvement locally one also had to make it locally.
+
+A more modern adoption model would obsessively make these improvements in
```
Comment (ChristianMurphy) [inline _posts/uportal/2017-12-06-myuw-fork.md]: "obsessively" sounds like a negatively loaded term, is that intended?

File: _posts/uportal/2017-12-06-myuw-fork.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
++ Removed `x:SearchRequest` and `x:SearchResponse` event handling in 
+`SimpleContentPortlet` `portlet.xml`
++ Removed jGroups cache invalidation signaling and listening
++ In process of removing Categories, use just one giant category that conveys 
+no permissions, assign all SUBSCRIBE and BROWSE permissions on a 
+per-portlet-definition basis
+
+# Other fixes
+
++ Fix deleting permission grants via permissions management Portlet
```
Comment (ChristianMurphy) [inline _posts/uportal/2017-12-06-myuw-fork.md]: Maybe rephrase? This could be ambiguous for folks that don't have in depth knowledge of portals permission system.

File: _posts/uportal/2017-12-06-myuw-fork.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+A more modern adoption model would obsessively make these improvements in 
+upstream rather than locally and then locally adopt them by upgrading to open 
+source product releases with the improvements.
+
+# The modifications
+
+## Hacked rendering pipeline
+
+MyUW configures the uPortal rendering pipeline to direct to
```
Comment (ChristianMurphy) [inline _posts/uportal/2017-12-06-myuw-fork.md]: This is supported upstream https://github.com/Jasig/uPortal/pull/436, https://github.com/Jasig/uPortal/pull/708,  https://github.com/Jasig/uPortal/pull/1050

File: _posts/uportal/2017-12-06-myuw-fork.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
++ Fewer regions. Simpler. Remove features MyUW isn't using.
++ When rendering a maximized portlet, simulate the way uPortal-home looks, so that rendering maximized portlets feels seamless enough. That is, stylize the header.
++ On add-to-home, clear uPortal-home-specific session storage items so that change will reflect when rendering `uPortal-home`
+
+
+# Hacked JSON APIs
+
+MyUW has modified JSON APIs to expose data `uPortal-home` needs.
+
++ Expose more data about portlet publications, to power widgets. e.g. `renderOnWeb`,`mdIcon`, `widgetURL`, `widgetType`, `widgetConfig`, `widgetTemplate`, `widgetExternalMessageUrl`, `widgetExternalMesageTextObjectLocation`, `widgetExternalMessageLearnMoreUrlLocation`, in `LayoutPortlet`. See also idea to use portlet publishing parameters for this instead.
```
Comment (ChristianMurphy) [inline _posts/uportal/2017-12-06-myuw-fork.md]: > See also idea to use portlet publishing parameters for this instead.

:link: https://github.com/uPortal-Project/uportal-home/issues/753

File: _posts/uportal/2017-12-06-myuw-fork.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
++ Add `/portlet/{fname}.json` to get JSON for a single portlet
++ Add `GroupRESTController`
++ Defaulting `url` of LayoutPortlet to that to render the portlet via uPortal
++ Add `lifecycleState` to `LayoutPortlet`
++ Add simplified move-portlet API call (`action=movePortletAjax` in `UpdatePreferencesServlet`)
++ Add remove-from-layout-by-fname (`action=removeByFName` in `UpdatePreferencesServlet`)
++ Add `firstName`, `lastName` in the `SessionRESTController` response. Also, `originalUsername`  when swapped.
++ Add some error handling in `LayoutPortlet` to catch when `widgetConfig` is not valid JSON.
++ Increase time-to-live of `marketplacePortletDefinitionCache`
++ Filter marketplace entries to only those in `PUBLISHED` publication lifecycle state. (Arguably this should be more nuanced, reflecting the viewing user's permissions -- maybe you've got permission to see entries that are in states other than `PUBLISHED`).
```
Comment (ChristianMurphy) [inline _posts/uportal/2017-12-06-myuw-fork.md]: Related to https://github.com/Jasig/uPortal/pull/1068

File: _posts/uportal/2017-12-06-myuw-fork.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
++ Attempts to make hitting JSON APIs have the side effect of logging the user in / bootstrapping a uPortal session when they can. This is fraught.
+
+# Hacked DLM
+
+MyUW has modified DLM to simplify the user experience of adding content to one's homepage.
+
++ Default adding a new item to the first rather than last position in a container.
```
Comment (ChristianMurphy) [inline _posts/uportal/2017-12-06-myuw-fork.md]: There was an effort to upstream this in https://github.com/Jasig/uPortal/pull/885 the PR stalled before it could be merged. :disappointed:

File: _posts/uportal/2017-12-06-myuw-fork.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
+---
+layout:     post
+title:      MyUW uPortal fork: the good parts
```
Comment (ChristianMurphy) [inline _posts/uportal/2017-12-06-myuw-fork.md]: I'm cautious about phrasing here.
The article is about features and optimization made by MyUW's portal, changes which I hope to see upstreamed or reconciled with Apereo uPortal.
This title could be misconstrued as advocating for forking Apereo uPortal, which I don't think is the article's goal.

File: _posts/uportal/2017-12-06-myuw-fork.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
++ Prefer to render static content via `/web/static/{fname}`
++ Prefer to render portlets via `/web/exclusive/{fname}` (`renderOnWeb`)
++ Bounce attempts to do anything other than render just one portlet (i.e. maximized or exclusive) to `/web`
++ Coerce guest user to the desired profile
+
+# Hacked theme
+
+MyUW uses a customized "Bucky" theme forked from Respondr. The purpose of this customized theme and associated skins is to make rendering a portlet maximized via uPortal look like rendering uPortal-home, so that users switch reasonably seamlessly between the two approaches depending on content.
+
++ Fewer regions. Simpler. Remove features MyUW isn't using.
```
Comment (ChristianMurphy) [inline _posts/uportal/2017-12-06-myuw-fork.md]: uPortal 5 has changed a number of regions to portlets that can easily be removed. E.G. https://github.com/Jasig/uPortal/pull/1081

File: _posts/uportal/2017-12-06-myuw-fork.md [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+Prompt the user to log out after resetting layout.
+
+Fix non-initialized field in GroupsExceptionImpl.
+
+# Microservices
+
++ Rest Proxy
++ Key-value store
++ LTI Redirector
++ Personalized Redirector
```
Comment (ChristianMurphy) [inline _posts/uportal/2017-12-06-myuw-fork.md]: Do these services rely on a forked version of the portal?

**Comment by apetro** [comment]
Still intend to circle back on this.
